### ISSUE TITLE
Redirecting banned HTTP users to another webserver

### ISSUE THREAD

**Issue by szepeviktor** [fail2ban/fail2ban#1113]
Would you merge a "*Friendly WAF*" action?

```ini
actionban = iptables -t nat -A PREROUTING -i <interface> -p tcp -s <ip> --dport <webport> -j REDIRECT --to-port <wafport>

interface = eth0
webport = 80
wafport = 8080
```
And users *banned* from HTTP will see only one HTML page (with dependencies on an external host/CDN).

Could you suggest a light webserver for 1 static HTML page?

**Comment by buanzo** [comment]
Python httpserver module.

**Comment by leeclemens** [comment]
I don't like the idea of fail2ban being exposed directly to external users, especially seeing as it runs as root.

Having fail2ban manage an iptables rule to REDIRECT makes sense though, is that what you're asking about?

If you're just asking about a webserver recommendation, this isn't the place. Feel free to reopen if you have a question regarding fail2ban (or better yet, ask the mailing list).

**Comment by yarikoptic** [comment]
Where does fail2ban get exposed here? It will not provide webserver, just a new action to redirect. IMHO worthwhile addition


On July 12, 2015 8:30:47 AM EDT, Lee Clemens <notifications@github.com> wrote:
>I don't like the idea of fail2ban being exposed directly to external
>users, especially seeing as it runs as root.
>
>Having fail2ban manage an iptables rule to REDIRECT makes sense though,
>is that what you're asking about?
>
>If you're just asking about a webserver recommendation, this isn't the
>place. Feel free to reopen if you have a question regarding fail2ban
>(or better yet, ask the mailing list).
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/fail2ban/fail2ban/issues/1113#issuecomment-120717949

-- 
Sent from a phone which beats iPhone.

**Comment by leeclemens** [comment]
> Having fail2ban manage an iptables rule to REDIRECT makes sense though, is that what you're asking about?

Having fail2ban act as that webserver in any way would expose it. I read the initial post as being related to fail2ban, beyond having fail2ban create the REDIRECT in iptables (see above).

**Comment by buanzo** [comment]
I understood that Viktor was asking for a way to REDIRECT web users to
another port, (action), and was also asking for a suggestion on a very
low-resource-usage http server. I think 'friendly waf' is useful to have in
f2b, and think like Yaroslav on this one.

**Comment by leeclemens** [comment]
Maybe the three of you can help me understand the term "[Friendly WAF] (https://www.google.com/search?q=%22Friendly+WAF%22)".  I don't think general questions for web server recommendations are directly appropriate as a fail2ban Issue, and I'd hate to see fail2ban try to become one in any way (read: exposing fail2ban).

@yarikoptic To explain further and without trying it out myself, I think there are enough variables already to accomplish the REDIRECT approach (between `<chain>` and `<blocktype>`) but we'd probably have to add `-t nat` (as a variable, obviously). Then it would fit in nicely without sprawl. Why create a new specific action for one case?

**Comment by leeclemens** [comment]
@buanzo I too agree with fail2ban having the ability to REDIRECT...so I think we're all on the same page.

**Comment by szepeviktor** [comment]
A friendly WAF tells you that you are banned. See CloudFlare's "blocked" page.
Now a banned HTTP user gets a connection failed browser message, or something like that.

**Comment by leeclemens** [comment]
ok, I stand by my initial response that fail2ban should not serve a "friendly waf" page, as that would require fail2ban to accept and respond to http requests from external sources.

**Comment by szepeviktor** [comment]
I am sorry for writing so short comment. Of course fail2ban should only REDIRECT.
I am experimenting with [lighttpd2](https://github.com/lighttpd/lighttpd2) as the webserver must be able to server many virtual hosts and TLS also.

**Comment by szepeviktor** [comment]
> Could you suggest a light webserver for 1 static HTML page?

This question is there only because I don't know any lightweight webserver.

**Comment by leeclemens** [comment]
I think you could accomplish the same goal using Apache, or whatever http server you use, and that the discussion of a webserver recommendation isn't really relevant to fail2ban's Issue forum (I used "direct" before intentionally). That being said...it seems you just want fail2ban to create the REDIRECT iptables rule, right? If so, seems we can circle back to https://github.com/fail2ban/fail2ban/issues/1113#issuecomment-120730933

**Comment by szepeviktor** [comment]
So: I'd like to make make fail2ban + lighthttpd2 a friendly WAF displaying a "banned" message to users.

**Comment by szepeviktor** [comment]
@leeclemens Please suggest a name for the new action.

**Comment by leeclemens** [comment]
@szepeviktor As I mentioned, I don't think this needs a new action, as it should be able to be accomplished using the existing iptables-multiport action with the addition of `-t <table>`. Sprawl is bad.

**Comment by szepeviktor** [comment]
It is beyond my capabilities to get this
`iptables -t nat -D PREROUTING -i <interface> -p tcp -s <ip> --dport <webport> -j REDIRECT --to-port <wafport>`
from the existing action:
start: `iptables -I <chain> -p <protocol> -m multiport --dports <port> -j f2b-<name>`
ban: `iptables -I f2b-<name> 1 -s <ip> -j <blocktype>`

**Comment by buanzo** [comment]
I think nobody suggested that, Lee.

**Comment by leeclemens** [comment]
@buanzo thanks...please check the definition of "addition". If you disagree with my statement that it shouldn't be a new action, please response. If not, I'm not sure what value you think your last comment has.

@szepeviktor If that's the case, fine. You initially asked
 > Would you merge [...]

All of my responses are consistent with that. Moving on, to use iptables-multiport for this new application shouldn't be too difficult. You already have a PR making changes to iptables actions and adding the variable `<returntype>`. I think we just need another `-t <tablename>` default to `filter`, so you can change it to `nat`. `<blocktype>` is already a variable...you'd just need to override it to `-j REDIRECT --to-port 8080` for your custom application/whatever. Maybe a variable for `redirectport` for icing on the cake. You've already made changes to iptables configs, I'm just trying to point you in the right direction in order to implement this new enhancement

**Comment by szepeviktor** [comment]
You are very kind.

**Comment by buanzo** [comment]
I did not comment on this thread.

**Comment by szepeviktor** [comment]
https://github.com/fail2ban/fail2ban/issues/1113#issuecomment-120751135 ?

**Comment by leeclemens** [comment]
@buanzo Please elaborate on what you meant by "that" in https://github.com/fail2ban/fail2ban/issues/1113#issuecomment-120751135?

**Comment by yarikoptic** [comment]
Indeed! May be it could already be achieved by modifying blocktype variable and adding to it -t nat ? Then we would just need a jail spec for this feature

**Comment by leeclemens** [comment]
Need to add `-t whatever` to iptables-multport, defaults to `filter`, not `nat`, as we need it for the `REDIRECT`. I've probably spent more time discussing how to implement this better (imho) than it would have taken to implement it myself :(

**Comment by yarikoptic** [comment]
Technical discussions polish the end product! :-)

**Comment by leeclemens** [comment]
Agreed..."Update iptables-multport to specify table" or something and implemented hours ago :) Also "Would you merge" implies a forthcoming PR, not a feature request (and Issue to stay open until another PR is created).

**Comment by leeclemens** [comment]
this appears merged in https://github.com/fail2ban/fail2ban/pull/1112, should be closed.

**Comment by szepeviktor** [comment]
> this appears merged in #1112, should be closed.

Excuse me! This is also an issue opened by me.
But there is no `-j REDIRECT` in #1112.

**Comment by yarikoptic** [comment]
rright -- sorry, my bad... we are yet to cook up the full solution for this one.

**Comment by sebres** [comment]
> Please reopen this. I'll send the PR later.

Still interested?

**Comment by szepeviktor** [comment]
I was planning to mimic CloudFlare where suspicious traffic get a "You are blocked!" page and the chance to unblock himself.
